File: scripts/providers.py
"""LLM providers: Ollama native API and OpenAI-compatible APIs (OpenRouter, OpenAI,
LM Studio, llama.cpp, vLLM, ...). Standard library only."""

from __future__ import annotations

import json
import re
import time
import urllib.error
import urllib.request
from dataclasses import dataclass

from common import Config, config_problem, endpoint, is_local_endpoint, resolve_api_key, state_dir

LOCAL_HEALTH_TTL = 30
REMOTE_HEALTH_TTL = 600
RETRY_STATUSES = {408, 409, 429, 500, 502, 503, 504}
THINK_BLOCK = re.compile(r"<think>.*?</think>\s*", re.DOTALL)


class LLMError(RuntimeError):
    pass


@dataclass
class Health:
    ok: bool
    reason: str = ""


@dataclass
class ChatResult:
    text: str
    prompt_tokens: int | None
    output_tokens: int | None
    truncated: bool
    cost: float | None = None


class ModelOutputError(LLMError):
    """The request succeeded but the output is unusable; retrying will not help."""


class HTTPStatusError(LLMError):
    def __init__(self, status: int, message: str, retry_after: float | None = None):
        super().__init__(f"HTTP {status}: {message}")
        self.status = status
        self.retry_after = retry_after


def _error_message(body: str) -> str:
    try:
        data = json.loads(body)
    except ValueError:
        return body.strip()[:500]
    error = data.get("error", data) if isinstance(data, dict) else data
    if isinstance(error, dict):
        message = error.get("message") or json.dumps(error)
        # OpenRouter puts the upstream provider's explanation in metadata.
        metadata = error.get("metadata") if isinstance(error.get("metadata"), dict) else {}
        detail = metadata.get("raw") or metadata.get("reason")
        if detail:
            message = f"{message} ({metadata.get('provider_name') or 'upstream'}: {str(detail)[:300]})"
        error = message
    return str(error)[:500]


def http_json(url: str, payload: dict | None, timeout: float, headers: dict | None = None) -> dict:
    data = json.dumps(payload).encode() if payload is not None else None
    req = urllib.request.Request(url, data=data, headers={"Content-Type": "application/json", **(headers or {})})
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            result = json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        retry_after = e.headers.get("Retry-After") if e.headers else None
        try:
            retry_after = float(retry_after) if retry_after else None
        except ValueError:
            retry_after = None
        raise HTTPStatusError(e.code, _error_message(e.read().decode("utf-8", "replace")), retry_after) from e
    except (urllib.error.URLError, TimeoutError, OSError) as e:
        # Strip the query string; it never carries secrets here, but keep messages short.
        raise LLMError(f"cannot reach {url.split('?', 1)[0]}: {getattr(e, 'reason', e)}") from e
    except ValueError as e:
        raise LLMError(f"invalid JSON from {url}: {e}") from e
    if isinstance(result, dict) and result.get("error"):
        raise LLMError(f"API error: {_error_message(json.dumps(result))}")
    return result


# Tests set this to 0 to avoid sleeping.
BACKOFF_SCALE = 1.0


def backoff(attempt: int) -> float:
    """Delay before retry number attempt+1: 2, 4, 8 ... seconds, capped at 30."""
    return min(2 ** (attempt + 1), 30) * BACKOFF_SCALE


class Provider:
    name = ""

    def __init__(self, cfg: Config, cwd: str | None = None):
        self.cfg = cfg
        self.cwd = cwd
        self.base = endpoint(cfg)
        self.is_local = is_local_endpoint(self.base)

    # Subclasses implement these two.
    def list_models(self, timeout: float) -> list[str] | None: ...
    def _chat_once(self, system: str, user: str, max_output: int) -> ChatResult: ...

    def chat(self, system: str, user: str, max_output: int) -> ChatResult:
        attempts = 1 if self.is_local else 1 + max(0, self.cfg.max_retries)
        for attempt in range(attempts):
            try:
                result = self._chat_once(system, user, max_output)
                result.text = THINK_BLOCK.sub("", result.text or "").strip()
                return result
            except HTTPStatusError as e:
                if e.status not in RETRY_STATUSES or attempt == attempts - 1:
                    raise
                delay = e.retry_after * BACKOFF_SCALE if e.retry_after is not None and e.retry_after <= 60 else backoff(attempt)
            except LLMError as e:
                if isinstance(e, ModelOutputError) or attempt == attempts - 1:
                    raise
                delay = backoff(attempt)
            time.sleep(delay)
        raise LLMError("unreachable")


class OllamaProvider(Provider):
    name = "ollama"

    def list_models(self, timeout: float) -> list[str] | None:
        tags = http_json(f"{self.base}/api/tags", None, timeout)
        return [m.get("name", "") for m in tags.get("models", [])]

    def _chat_once(self, system: str, user: str, max_output: int) -> ChatResult:
        cfg = self.cfg
        payload = {
            "model": cfg.model,
            "messages": [{"role": "system", "content": system}, {"role": "user", "content": user}],
            "stream": False,
            "think": False,
            "keep_alive": cfg.keep_alive,
            "options": {"temperature": cfg.temperature, "num_ctx": cfg.num_ctx, "num_predict": max_output},
        }
        url = f"{self.base}/api/chat"
        try:
            resp = http_json(url, payload, cfg.request_timeout)
        except HTTPStatusError as e:
            # Some models reject the `think` field entirely; retry without it.
            if "think" not in str(e).lower():
                raise
            payload.pop("think")
            resp = http_json(url, payload, cfg.request_timeout)
        return ChatResult(
            text=resp.get("message", {}).get("content") or "",
            prompt_tokens=resp.get("prompt_eval_count"),
            output_tokens=resp.get("eval_count"),
            truncated=resp.get("done_reason") == "length",
        )


class OpenAICompatibleProvider(Provider):
    name = "openai-compatible"

    def __init__(self, cfg: Config, cwd: str | None = None):
        super().__init__(cfg, cwd)
        self.send_reasoning_off = cfg.disable_reasoning

    def headers(self) -> dict:
        headers = {}
        key = resolve_api_key(self.cfg, self.cwd)
        if key:
            headers["Authorization"] = f"Bearer {key}"
        if self.cfg.provider == "openrouter":
            headers["X-Title"] = "local-shunt"
        return headers

    def list_models(self, timeout: float) -> list[str] | None:
        try:
            data = http_json(f"{self.base}/models", None, timeout, self.headers())
        except HTTPStatusError as e:
            if e.status in (404, 405):
                return None  # server has no model listing; cannot verify
            raise
        return [m.get("id", "") for m in data.get("data", []) if isinstance(m, dict)]

    def reasoning_off_fields(self) -> dict:
        if self.cfg.provider == "openrouter":
            return {"reasoning": {"effort": "none"}}
        return {"reasoning_effort": "none"}

    def _chat_once(self, system: str, user: str, max_output: int) -> ChatResult:
        cfg = self.cfg
        payload = {
            "model": cfg.model,
            "messages": [{"role": "system", "content": system}, {"role": "user", "content": user}],
            "temperature": cfg.temperature,
            "max_tokens": max_output,
            "stream": False,
        }
        if self.send_reasoning_off:
            payload.update(self.reasoning_off_fields())
        payload.update(cfg.extra_body)
        url = f"{self.base}/chat/completions"
        try:
            resp = http_json(url, payload, cfg.request_timeout, self.headers())
        except HTTPStatusError as e:
            # Models without adjustable reasoning may reject the field; retry without it.
            if not (self.send_reasoning_off and e.status in (400, 422)):
                raise
            self.send_reasoning_off = False
            for key in self.reasoning_off_fields():
                if key not in cfg.extra_body:
                    payload.pop(key, None)
            resp = http_json(url, payload, cfg.request_timeout, self.headers())

        choices = resp.get("choices") or []
        if not choices:
            raise ModelOutputError("API returned no choices")
        choice = choices[0]
        message = choice.get("message") or {}
        usage = resp.get("usage") or {}
        text = message.get("content") or ""
        truncated = choice.get("finish_reason") == "length"
        if truncated and not text.strip() and (message.get("reasoning") or message.get("reasoning_content")):
            raise ModelOutputError(
                "the model used the whole output budget on reasoning and returned no answer; "
                "choose a non-reasoning model or raise --max-output"
            )
        return ChatResult(
            text=text,
            prompt_tokens=usage.get("prompt_tokens"),
            output_tokens=usage.get("completion_tokens"),
            truncated=truncated,
            cost=usage.get("cost") if isinstance(usage.get("cost"), (int, float)) else None,
        )


def get_provider(cfg: Config, cwd: str | None = None) -> Provider:
    problem = config_problem(cfg, cwd)
    if problem:
        raise LLMError(problem)
    if cfg.provider == "ollama":
        return OllamaProvider(cfg, cwd)
    return OpenAICompatibleProvider(cfg, cwd)


def model_available(model: str, names: list[str]) -> bool:
    if model in names:
        return True
    if ":" not in model:
        return f"{model}:latest" in names
    return False


def check_health(
    cfg: Config,
    timeout_ms: int | None = None,
    use_cache: bool = True,
    probe_remote: bool = False,
    cwd: str | None = None,
) -> Health:
    """Check that the provider is configured, reachable and has the model.

    Local endpoints are probed over HTTP. Remote endpoints are probed only when
    probe_remote is True (SessionStart); the hook relies on the cached result or,
    without one, on the configuration check alone, so it never waits on the network.
    """
    problem = config_problem(cfg, cwd)
    if problem:
        return Health(ok=False, reason=problem)

    base = endpoint(cfg)
    local = is_local_endpoint(base)
    ttl = LOCAL_HEALTH_TTL if local else REMOTE_HEALTH_TTL
    cache_file = state_dir() / "health.json"
    key = f"{cfg.provider}|{base}|{cfg.model}"
    if use_cache:
        try:
            cached = json.loads(cache_file.read_text(encoding="utf-8"))
            if cached.get("key") == key and time.time() - cached.get("ts", 0) < ttl:
                return Health(ok=cached["ok"], reason=cached.get("reason", ""))
        except (OSError, ValueError, KeyError):
            pass
    if not local and not probe_remote:
        return Health(ok=True, reason="remote endpoint not probed")

    timeout = (timeout_ms if timeout_ms is not None else cfg.health_timeout_ms) / 1000
    provider = get_provider(cfg, cwd)
    try:
        names = provider.list_models(timeout)
        if names is None or model_available(cfg.model, names):
            health = Health(ok=True)
        else:
            health = Health(ok=False, reason=f"model {cfg.model} is not available from {cfg.provider}")
    except LLMError as e:
        health = Health(ok=False, reason=str(e))

    try:
        cache_file.parent.mkdir(parents=True, exist_ok=True)
        cache_file.write_text(
            json.dumps({"key": key, "ts": time.time(), "ok": health.ok, "reason": health.reason}),
            encoding="utf-8",
        )
    except OSError:
        pass
    return health

File: internal/shunt/providers.go
package shunt

// LLM providers: Ollama's native API and OpenAI-compatible APIs (OpenRouter, OpenAI,
// LM Studio, llama.cpp, vLLM, ...).

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"
)

const (
	localHealthTTL  = 30
	remoteHealthTTL = 600
)

var (
	retryStatuses = map[int]bool{408: true, 409: true, 429: true, 500: true, 502: true, 503: true, 504: true}
	thinkBlock    = regexp.MustCompile(`(?s)<think>.*?</think>\s*`)
)

// Error kinds. They appear in the log as "error:<kind>".
const (
	kindLLM         = "LLMError"
	kindModelOutput = "ModelOutputError" // the request succeeded but the output is unusable; retrying will not help
	kindTimeout     = "RequestTimeout"   // retried like any transient remote failure (a one-off network stall isn't a slow model)
	kindHTTPStatus  = "HTTPStatusError"
)

// LLMError is a failure of the worker model or its API.
type LLMError struct {
	Kind       string
	Msg        string
	Status     int
	RetryAfter *float64
}

func (e *LLMError) Error() string {
	if e.Kind == kindHTTPStatus {
		return fmt.Sprintf("HTTP %d: %s", e.Status, e.Msg)
	}
	return e.Msg
}

func llmError(kind, format string, args ...any) *LLMError {
	return &LLMError{Kind: kind, Msg: fmt.Sprintf(format, args...)}
}

// Health is the result of a provider health check.
type Health struct {
	OK     bool
	Reason string
}

// ChatResult is one model response.
type ChatResult struct {
	Text         string
	PromptTokens *int
	OutputTokens *int
	Truncated    bool
	Cost         *float64
}

func truncateRunes(s string, n int) string {
	runes := []rune(s)
	if len(runes) <= n {
		return s
	}
	return string(runes[:n])
}

func errorMessage(body string) string {
	var data any
	if json.Unmarshal([]byte(body), &data) != nil {
		return truncateRunes(strings.TrimSpace(body), 500)
	}
	errValue := data
	if obj, ok := data.(map[string]any); ok {
		if inner, ok := obj["error"]; ok {
			errValue = inner
		}
	}
	if obj, ok := errValue.(map[string]any); ok {
		message := asString(obj["message"])
		if message == "" {
			raw, _ := marshalJSON(obj)
			message = string(raw)
		}
		// OpenRouter puts the upstream provider's explanation in metadata.
		metadata := asMap(obj["metadata"])
		detail := metadata["raw"]
		if !truthy(detail) {
			detail = metadata["reason"]
		}
		if truthy(detail) {
			name := asString(metadata["provider_name"])
			if name == "" {
				name = "upstream"
			}
			message = fmt.Sprintf("%s (%s: %s)", message, name, truncateRunes(pyStr(detail), 300))
		}
		return truncateRunes(message, 500)
	}
	return truncateRunes(pyStr(errValue), 500)
}

func stripQuery(url string) string {
	before, _, _ := strings.Cut(url, "?")
	return before
}

func httpJSON(url string, payload any, timeout time.Duration, headers map[string]string) (map[string]any, error) {
	method, body := http.MethodGet, io.Reader(nil)
	if payload != nil {
		data, err := marshalJSON(payload)
		if err != nil {
			return nil, llmError(kindLLM, "cannot encode request: %v", err)
		}
		method, body = http.MethodPost, bytes.NewReader(data)
	}
	req, err := http.NewRequest(method, url, body)
	if err != nil {
		return nil, llmError(kindLLM, "cannot reach %s: %v", stripQuery(url), err)
	}
	req.Header.Set("Content-Type", "application/json")
	for key, value := range headers {
		req.Header.Set(key, value)
	}
	client := &http.Client{Timeout: timeout}
	resp, err := client.Do(req)
	if err != nil {
		var netErr net.Error
		if errors.As(err, &netErr) && netErr.Timeout() {
			return nil, llmError(kindTimeout, "%s did not respond within %s s", stripQuery(url), strconv.FormatFloat(timeout.Seconds(), 'g', -1, 64))
		}
		reason := err
		var urlErr interface{ Unwrap() error }
		if errors.As(err, &urlErr) && urlErr.Unwrap() != nil {
			reason = urlErr.Unwrap()
		}
		return nil, llmError(kindLLM, "cannot reach %s: %v", stripQuery(url), reason)
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		var netErr net.Error
		if errors.As(err, &netErr) && netErr.Timeout() {
			return nil, llmError(kindTimeout, "%s did not respond within %s s", stripQuery(url), strconv.FormatFloat(timeout.Seconds(), 'g', -1, 64))
		}
		return nil, llmError(kindLLM, "cannot reach %s: %v", stripQuery(url), err)
	}
	if resp.StatusCode >= 400 {
		e := &LLMError{Kind: kindHTTPStatus, Status: resp.StatusCode, Msg: errorMessage(strings.ToValidUTF8(string(raw), "\uFFFD"))}
		if value := resp.Header.Get("Retry-After"); value != "" {
			if seconds, err := strconv.ParseFloat(strings.TrimSpace(value), 64); err == nil {
				e.RetryAfter = &seconds
			}
		}
		return nil, e
	}
	var result any
	if err := json.Unmarshal(raw, &result); err != nil {
		return nil, llmError(kindLLM, "invalid JSON from %s: %v", url, err)
	}
	obj, ok := result.(map[string]any)
	if !ok {
		return nil, llmError(kindLLM, "invalid JSON from %s: expected an object", url)
	}
	if truthy(obj["error"]) {
		data, _ := marshalJSON(obj)
		return nil, llmError(kindLLM, "API error: %s", errorMessage(string(data)))
	}
	return obj, nil
}

// sleep is replaced in tests to avoid waiting.
var sleep = time.Sleep

// backoff is the delay before retry number attempt+1: 2, 4, 8 ... seconds, capped at 30.
func backoff(attempt int) time.Duration {
	return time.Duration(math.Min(math.Pow(2, float64(attempt+1)), 30) * float64(time.Second))
}

// Provider talks to one worker model endpoint.
type Provider struct {
	Cfg              Config
	cwd              string
	base             string
	IsLocal          bool
	sendReasoningOff bool
}

// getProvider validates the configuration and returns a provider for it.
func getProvider(cfg Config, cwd string) (*Provider, error) {
	if problem := configProblem(cfg, cwd); problem != "" {
		return nil, llmError(kindLLM, "%s", problem)
	}
	base := endpoint(cfg)
	return &Provider{Cfg: cfg, cwd: cwd, base: base, IsLocal: isLocalEndpoint(base), sendReasoningOff: cfg.DisableReasoning}, nil
}

func (p *Provider) ollama() bool { return p.Cfg.Provider == "ollama" }

func (p *Provider) requestTimeout() time.Duration {
	return time.Duration(p.Cfg.RequestTimeout) * time.Second
}

// Chat sends one system and user message, retrying transient failures of remote endpoints.
func (p *Provider) Chat(system, user string, maxOutput int) (ChatResult, error) {
	attempts := 1
	if !p.IsLocal {
		attempts += max(0, p.Cfg.MaxRetries)
	}
	for attempt := 0; ; attempt++ {
		var (
			result ChatResult
			err    error
		)
		if p.ollama() {
			result, err = p.ollamaChat(system, user, maxOutput)
		} else {
			result, err = p.compatChat(system, user, maxOutput)
		}
		if err == nil {
			result.Text = strings.TrimSpace(thinkBlock.ReplaceAllString(result.Text, ""))
			return result, nil
		}
		var e *LLMError
		if !errors.As(err, &e) || attempt == attempts-1 {
			return result, err
		}
		delay := backoff(attempt)
		switch e.Kind {
		case kindHTTPStatus:
			if !retryStatuses[e.Status] {
				return result, err
			}
			if e.RetryAfter != nil && *e.RetryAfter <= 60 {
				delay = time.Duration(*e.RetryAfter * float64(time.Second))
			}
		case kindModelOutput:
			return result, err
		}
		sleep(delay)
	}
}

// listModels returns the model names, or nil with ok=false when the server has no listing.
func (p *Provider) listModels(timeout time.Duration) (names []string, ok bool, err error) {
	if p.ollama() {
		tags, err := httpJSON(p.base+"/api/tags", nil, timeout, nil)
		if err != nil {
			return nil, false, err
		}
		models, _ := tags["models"].([]any)
		for _, m := range models {
			names = append(names, asString(asMap(m)["name"]))
		}
		return names, true, nil
	}
	data, err := httpJSON(p.base+"/models", nil, timeout, p.headers())
	if err != nil {
		var e *LLMError
		if errors.As(err, &e) && e.Kind == kindHTTPStatus && (e.Status == 404 || e.Status == 405) {
			return nil, false, nil // server has no model listing; cannot verify
		}
		return nil, false, err
	}
	models, ok := data["data"].([]any)
	if !ok {
		return nil, false, nil // response doesn't match the expected shape; cannot verify
	}
	for _, m := range models {
		if obj, isObj := m.(map[string]any); isObj {
			names = append(names, asString(obj["id"]))
		}
	}
	return names, true, nil
}

func (p *Provider) ollamaChat(system, user string, maxOutput int) (ChatResult, error) {
	cfg := p.Cfg
	payload := map[string]any{
		"model":      cfg.Model,
		"messages":   []map[string]string{{"role": "system", "content": system}, {"role": "user", "content": user}},
		"stream":     false,
		"think":      false,
		"keep_alive": cfg.KeepAlive,
		"options":    map[string]any{"temperature": cfg.Temperature, "num_ctx": cfg.NumCtx, "num_predict": maxOutput},
	}
	url := p.base + "/api/chat"
	resp, err := httpJSON(url, payload, p.requestTimeout(), nil)
	var e *LLMError
	if err != nil && errors.As(err, &e) && e.Kind == kindHTTPStatus && strings.Contains(strings.ToLower(e.Error()), "think") {
		// Some models reject the `think` field entirely; retry without it.
		delete(payload, "think")
		resp, err = httpJSON(url, payload, p.requestTimeout(), nil)
	}
	if err != nil {
		return ChatResult{}, err
	}
	return ChatResult{
		Text:         asString(asMap(resp["message"])["content"]),
		PromptTokens: intPtr(resp["prompt_eval_count"]),
		OutputTokens: intPtr(resp["eval_count"]),
		Truncated:    resp["done_reason"] == "length",
	}, nil
}

func (p *Provider) headers() map[string]string {
	headers := map[string]string{}
	if key := resolveAPIKey(p.Cfg, p.cwd); key != "" {
		headers["Authorization"] = "Bearer " + key
	}
	if p.Cfg.Provider == "openrouter" {
		headers["X-Title"] = "local-shunt"
	}
	return headers
}

func (p *Provider) reasoningOffFields() map[string]any {
	if p.Cfg.Provider == "openrouter" {
		return map[string]any{"reasoning": map[string]any{"effort": "none"}}
	}
	return map[string]any{"reasoning_effort": "none"}
}

func (p *Provider) compatChat(system, user string, maxOutput int) (ChatResult, error) {
	cfg := p.Cfg
	payload := map[string]any{
		"model":       cfg.Model,
		"messages":    []map[string]string{{"role": "system", "content": system}, {"role": "user", "content": user}},
		"temperature": cfg.Temperature,
		"max_tokens":  maxOutput,
		"stream":      false,
	}
	if p.sendReasoningOff {
		for key, value := range p.reasoningOffFields() {
			payload[key] = value
		}
	}
	for key, value := range cfg.ExtraBody {
		payload[key] = value
	}
	url := p.base + "/chat/completions"
	resp, err := httpJSON(url, payload, p.requestTimeout(), p.headers())
	var e *LLMError
	if err != nil && p.sendReasoningOff && errors.As(err, &e) && e.Kind == kindHTTPStatus && (e.Status == 400 || e.Status == 422) {
		// Models without adjustable reasoning may reject the field; retry without it.
		p.sendReasoningOff = false
		for key := range p.reasoningOffFields() {
			if _, set := cfg.ExtraBody[key]; !set {
				delete(payload, key)
			}
		}
		resp, err = httpJSON(url, payload, p.requestTimeout(), p.headers())
	}
	if err != nil {
		return ChatResult{}, err
	}

	choices, _ := resp["choices"].([]any)
	if len(choices) == 0 {
		return ChatResult{}, llmError(kindModelOutput, "API returned no choices")
	}
	choice := asMap(choices[0])
	message := asMap(choice["message"])
	usage := asMap(resp["usage"])
	text := asString(message["content"])
	truncated := choice["finish_reason"] == "length"
	if truncated && strings.TrimSpace(text) == "" && (truthy(message["reasoning"]) || truthy(message["reasoning_content"])) {
		return ChatResult{}, llmError(kindModelOutput,
			"the model used the whole output budget on reasoning and returned no answer; "+
				"choose a non-reasoning model or raise --max-output")
	}
	result := ChatResult{
		Text:         text,
		PromptTokens: intPtr(usage["prompt_tokens"]),
		OutputTokens: intPtr(usage["completion_tokens"]),
		Truncated:    truncated,
	}
	if cost, ok := number(usage["cost"]); ok {
		result.Cost = &cost
	}
	return result, nil
}

func modelAvailable(model string, names []string) bool {
	for _, name := range names {
		if name == model || (!strings.Contains(model, ":") && name == model+":latest") {
			return true
		}
	}
	return false
}

// healthOptions tune checkHealth.
type healthOptions struct {
	timeoutMS   int // 0 uses health_timeout_ms
	noCache     bool
	probeRemote bool
	cwd         string
}

// checkHealth checks that the provider is configured, reachable and has the model.
//
// Local endpoints are probed over HTTP. Remote endpoints are probed only when
// probeRemote is set (SessionStart); the hook relies on the cached result or,
// without one, on the configuration check alone, so it never waits on the network.
func checkHealth(cfg Config, opts healthOptions) Health {
	if problem := configProblem(cfg, opts.cwd); problem != "" {
		return Health{Reason: problem}
	}

	base := endpoint(cfg)
	local := isLocalEndpoint(base)
	ttl := float64(remoteHealthTTL)
	if local {
		ttl = localHealthTTL
	}
	cacheFile := filepath.Join(stateDir(), "health.json")
	key := cfg.Provider + "|" + base + "|" + cfg.Model
	now := float64(time.Now().UnixNano()) / 1e9
	if !opts.noCache {
		if data, err := os.ReadFile(cacheFile); err == nil {
			var cached struct {
				Key    string  `json:"key"`
				TS     float64 `json:"ts"`
				OK     *bool   `json:"ok"`
				Reason string  `json:"reason"`
			}
			if json.Unmarshal(data, &cached) == nil && cached.OK != nil && cached.Key == key && now-cached.TS < ttl {
				return Health{OK: *cached.OK, Reason: cached.Reason}
			}
		}
	}
	if !local && !opts.probeRemote {
		return Health{OK: true, Reason: "remote endpoint not probed"}
	}

	timeoutMS := opts.timeoutMS
	if timeoutMS == 0 {
		timeoutMS = cfg.HealthTimeoutMS
	}
	provider, err := getProvider(cfg, opts.cwd)
	if err != nil {
		return Health{Reason: err.Error()}
	}
	var health Health
	names, listed, err := provider.listModels(time.Duration(timeoutMS) * time.Millisecond)
	switch {
	case err != nil:
		health = Health{Reason: err.Error()}
	case !listed || modelAvailable(cfg.Model, names):
		health = Health{OK: true}
	default:
		health = Health{Reason: fmt.Sprintf("model %s is not available from %s", cfg.Model, cfg.Provider)}
	}

	if os.MkdirAll(filepath.Dir(cacheFile), 0o755) == nil {
		data, _ := marshalJSON(record{{"key", key}, {"ts", now}, {"ok", health.OK}, {"reason", health.Reason}})
		_ = os.WriteFile(cacheFile, data, 0o644)
	}
	return health
}
